### ISSUE TITLE
Support request for model AMF765/10

### ISSUE THREAD

**Issue by sdimitrov-sms** [kongo09/philips-airpurifier-coap#63]
Hey there! We talked over the HASS forum, and here is what I found so far:

```
{
   "D01102":5,
   "D01S03":"Living Room Air Control",
   "D01S04":"Pegasus",
   "D01S05":"AMF765/10",
   "D01107":0,
   "D01108":3,
   "D01109":1,
   "D0110A":0,
   "D0110B":7,
   "D0110C":3,
   "D0110F":3,
   "D01S12":"0.8.5",
   "D01213":0,
   "D01R22":"0.1.5",
   "D01223":0,
   "D01R32":"0.3.0",
   "D01233":0,
   "ProductId":"5537ddae116411eda7a60217247a73aa",
   "DeviceId":"a2370e3699ae11ed9ecdefa402faee54",
   "MCUBoot":true,
   "Runtime":16768577,
   "rssi":-45,
   "wifilog":false,
   "free_memory":40832,
   "WifiVersion":"AWS_Philips_AIR@76",
   "StatusType":"status",
   "ConnectType":"Online",
   "D03102":1,
   "D03103":0,
   "D03106":0,
   "D0310A":2,
   "D0310C":0,
   "D0310D":1,
   "D0310E":25,
   "D0320F":30,
   "D03110":0,
   "D03113":0,
   "D03114":0,
   "D0311F":0,
   "D03120":1,
   "D03221":1,
   "D03122":1,
   "D03123":50,
   "D03224":250,
   "D03125":54,
   "D0312A":1,
   "D0312B":1,
   "D0312C":4,
   "D0312D":30,
   "D0312E":30,
   "D0312F":0,
   "D03130":100,
   "D03131":0,
   "D03134":0,
   "D03240":0,
   "D03180":1,
   "D03182":1,
   "D03R81":"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAA==\n",
   "D05102":0,
   "D05207":720,
   "D05408":14400,
   "D0520D":716,
   "D0540E":14400
}
```

**Comment by kongo09** [comment]
As discussed here: https://community.home-assistant.io/t/philips-air-purifier/53030/386?u=kongo09

**Comment by sdimitrov-sms** [comment]
I think those are the clear ones:

Field | Value | Mapping | Comment
------|-------|-----------|------------
  "D01102" | "5" | ?
  "D01S03" | "Living Room Air Control" | "name"
  "D01S04" | "Pegasus" | ? | 
  "D01S05" | "AC1715/11" | "modelid"
  "ProductId" | "5537ddae116411eda7a60217247a73aa" | ProductId
  "DeviceId" | "a2370e3699ae11ed9ecdefa402faee54" | DeviceId
  "MCUBoot" | true | ?
  "Runtime" | 16768577 | "Runtime"
  "rssi" | -65 | "rssi"
  "wifilog" | false | "wifilog"
  "free_memory" | 57544 | "free_memory"
  "WifiVersion" | "AWS_Philips_AIR@76" | "WifiVersion"
  "StatusType" | "status" | "StatusType"
  "ConnectType" | "Online" | "ConnectType"
  "D01S12" | "0.8.5" | -- | device firmware
  "D0320F" | 290 | -- | fan oscillation degrees - from 0 to 350
  "D0312E" | 70 | -- | screen light brigthness - from 0 to 100%, incrementals by 10
  "D0310E" | 25 | -- | indoor temperature?
  "D03125" | 53 | -- | humidity?

**Comment by kongo09** [comment]
Field | Value | Mapping | Comment
------|-------|-----------|------------
"D01102" | 5 |
"D01S03" | "Living Room Air Control" |
"D01S04" | "Pegasus" |
"D01S05" | "AMF765/10" |
"D01107" | 0 |
"D01108" | 3 |
"D01109" | 1 |
"D0110A" | 0 |
"D0110B" | 7 |
"D0110C" | 3 |
"D0110F" | 3 |
"D01S12" | "0.8.5" |
"D01213" | 0 |
"D01R22" | "0.1.5" |
"D01223" | 0 |
"D01R32" | "0.3.0" |
"D01233" | 0 |
"ProductId" | "5537ddae116411eda7a60217247a73aa" |
"DeviceId" | "a2370e3699ae11ed9ecdefa402faee54" |
"MCUBoot" | true |
"Runtime" | 16768577 |
"rssi" | -45 |
"wifilog" | false |
"free_memory" | 40832 |
"WifiVersion" | "AWS_Philips_AIR@76" |
"StatusType" | "status" |
"ConnectType" | "Online" |
"D03102" | 1 |
"D03103" | 0 |
"D03106" | 0 |
"D0310A" | 2 |
"D0310C" | 0 |
"D0310D" | 1 |
"D0310E" | 25 |
"D0320F" | 30 |
"D03110" | 0 |
"D03113" | 0 |
"D03114" | 0 |
"D0311F" | 0 |
"D03120" | 1 |
"D03221" | 1 |
"D03122" | 1 |
"D03123" | 50 |
"D03224" | 250 |
"D03125" | 54 |
"D0312A" | 1 |
"D0312B" | 1 |
"D0312C" | 4 |
"D0312D" | 30 |
"D0312E" | 30 |
"D0312F" | 0 |
"D03130" | 100 |
"D03131" | 0 |
"D03134" | 0 |
"D03240" | 0 |
"D03180" | 1 |
"D03182" | 1 |
"D03R81" | "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\nAAAAAAAAAAAAAA==\n" |
"D05102" | 0 |
"D05207" | 720 |
"D05408" | 14400 |
"D0520D" | 716 |
"D0540E" | 14400 |

**Comment by kongo09** [comment]
Thanks. We'll need the values for the different operating states. To get these, you should systematically change the modes in the app, get the status from the command line and figure out what has changed.

I suspect the values in the end `D05207` to `D0540E` are the filter time and filter max time. You should see them counting up.

**Comment by sdimitrov-sms** [comment]
I got some of the values in my previous comment. Tomorrow I will try to gather more data.

**Comment by kongo09** [comment]
try pre-release: `v0.11.4-beta765.1`

**Comment by sdimitrov-sms** [comment]
Hey @kongo09, just tried it, however, I am getting this error:

`Logger: aiohttp.server
Source: custom_components/philips_airpurifier_coap/config_flow.py:198
Integration: Philips AirPurifier ([documentation](https://github.com/kongo09/philips-airpurifier-coap), [issues](https://github.com/kongo09/philips-airpurifier-coap/issues))
First occurred: January 22, 2023 at 23:40:22 (3 occurrences)
Last logged: January 22, 2023 at 23:59:48

Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/aiohttp/web_protocol.py", line 435, in _handle_request
    resp = await request_handler(request)
  File "/usr/local/lib/python3.10/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/usr/local/lib/python3.10/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 60, in security_filter_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/forwarded.py", line 100, in forwarded_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 81, in ban_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 236, in auth_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 136, in handle
    result = await result
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 180, in post
    return await super().post(request, flow_id)
  File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 73, in wrapper
    result = await method(view, request, data, *args, **kwargs)
  File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 110, in post
    result = await self._flow_mgr.async_configure(flow_id, data)
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 286, in async_configure
    result = await self._async_handle_step(
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 374, in _async_handle_step
    result: FlowResult = await getattr(flow, method)(user_input)
  File "/config/custom_components/philips_airpurifier_coap/config_flow.py", line 198, in async_step_user
    self._model = list(filter(None, map(status.get, [PHILIPS_MODEL_ID, PHILIPS_NEW_MODEL_ID])))[0][:6]
IndexError: list index out of range`

**Comment by kongo09** [comment]
Ah, my mistake, it cannot work. Thanks for the log. I'll do an update soon.

Beware, the beta won't do much right now. It's only to get us started... Have you been able to get more data collected for the various states?

**Comment by sdimitrov-sms** [comment]
Hey again, not yet since I am swamped with work but soon I will get some details!

**Comment by kongo09** [comment]
try [v0.11.4-beta765.2](https://github.com/kongo09/philips-airpurifier-coap/releases/tag/v0.11.4-beta765.2)

still no features (waiting for your input), but recognition should work now

**Comment by grssll** [comment]
Tried to add same model AMF765/70 with no luck. Timeout, cannot connect to device.
v0.14

**Comment by codingcatgirl** [comment]
Got the thing today. `status` and `status-observer` worked not perfectly, but well enough to gather the data. Now, towards the end, i stopped getting data from them, it just gets stuck waiting for data.

Here are my findings:

| Key	        | Example Value	        | Seen values	| Description
| ------------- | ------------- | ------------- | ------------------------| 
| "D01102"	| 5		| always 5 	|
| "D01S03"	| "Living Room" |
| "D01S04"	| "Pegasus" 	|
| "D01S05"	| "AMF765/10" 	|
| "D01107"	| 0		| always 0 	|
| "D01108"	| 3		| always 3 	|
| "D01109"	| 1		| always 1 	|
| "D0110A"	| 0		| always 0 	|
| "D0110B"	| 7		| always 7 	|
| "D0110C"	| 3		| always 3 	|
| "D0110F"	| 3		| always 3 	|
| "D01S12"	| "0.8.6"	| 0.7.3, 0.8.6	| Device Firmware Version |
| "D01213"	| 0		| always 0 	|
| "D01R22"	| "0.1.5"	| always 0.1.5 	|
| "D01223"	| 0		| always 0 	|
| "D01R32"	| "0.3.0"	| always 0.3.0 	|
| "D01233"	| 0		| always 0 	|
| "ProductId"	| "5537ddae116411eda7a60217247a73aa" |
| "DeviceId"	| "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" |
| "MCUBoot"	| true		| always true 	|
| "Runtime"	| 701713	| 		| counting up, milliseconds |
| "rssi"	| -34		| 
| "wifilog"	| false		| always false 	|
| "free_memory"	| 47096		| 
| "WifiVersion"	| "AWS_Philips_AIR@78" | `@72.6`, `@78` | WiFi Firmware Version |
| "StatusType"	| "status"	| "status", "control" |
| "ConnectType"	| "Online"	| "Online", 	|
| "D03102"	| 1		| 0, 1		| Power status (0=standby, 1=on) |
| "D03103"	| 0		| 0, 1		| Child lock (0=off, 1=on) |
| "D03106"	| 0		| 0, 1 		| no idea what made it go from 0 to 1
| "D0310A"	| 2		| 1, 2		| Mode (1=fan, 2=circulation) (opens/closes extra vent) |
| "D0310C"	| 5		| 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 17, 18 | Fan setting (0=auto mode, 1-10=manual speed, 17=sleep mode, 18=turbo mode) – 0 during standby |
| "D0310D"	| 5		| 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 18 | Fan actual (1-10=speed, 18=turbo) – 0 during standby |
| "D0310E"	| 25		| always 25	| (not temperature)
| "D0320F"	| 0		| 0, 45, 50, 65, …, 350	| fan oscillation degrees |
| "D03110"	| 0		| 0-13		| poweroff timer (0=off, 1=0.5h, 2=1h, 3=2h, 4=3h, ..., 13=12h) |
| "D03113"	| 0		| always 0 	|
| "D03114"	| 0		| always 0 	|
| "D0311F"	| 0		| always 0 	|
| "D03120"	| 2		| 0, 1, 2	| unknown sensor value – 0 during standby if sensors are off, IAI maybe? but it's not shown in the App at all… |
| "D03221"	| 7		| 2-9		| PM2.5 – 0 during standby if sensors are off |
| "D03122"	| 1		| always 1 	|
| "D03123"	| 50		| always 50 	|
| "D03224"	| 236		| 238-249	| temperature in °C ×10 |
| "D03125"	| 57		| 53-57		| relative humidity % |
| "D0312A"	| 1		| always 1 	|
| "D0312B"	| 1		| always 1 	|
| "D0312C"	| 4		| always 4 	|
| "D0312D"	| 0		| 0, 10, …, 100	| display brightness (0=off after a few seconds) |
| "D0312E"	| 30		| always 30 	|
| "D0312F"	| 0		| always 0	|	 |
| "D03130"	| 100		| 0, 100	| beep volume |
| "D03131"	| 0		| always 0 	|
| "D03134"	| 0		| 0, 1		| sensors active during standby (0=no, 1=yes) |
| "D03240"	| 0		| always 0 	|
| "D03180"	| 0		| 0, 1		| Auto+ AI Mode (0=off, 1=yes) |
| "D03182"	| 1		| N/A, 1	| Only introduced in an update, wasn't there before, since then always 1 |
| "D03R81"	| "[endless screaming in base64]" |
| "D05102"	| 0		| always 0	|
| "D05207"	| 720		| always 720	|
| "D05408"	| 14400		| always 14400	|
| "D0520D"	| 720		| always 720	|
| "D0540E"	| 14400		| always 14400	|

I hope that you can add support with this. If you need any information to help debug/test out the pairing/connection issues (honestly, it feels like the app isn't doing very well either, but it's still getting data at least) i'm available to test out stuff.

One of these will probably also be the air quality level. Unfortunately, the air quality in here is too good and i didn't manage to change the light color.

**Comment by codingcatgirl** [comment]
Left it off over night to see if air quality gets worse, and for once i am happy about worse air quality – more interesting values!

I will post the new table i na second, but here's a summary of what changed:

- `"D03221"` (PM2.5 concentration) has risen to 21-22
- `"D03120"` (unknown sensor value)  has gone up to 4-5, confirming that it has a a bit of a wider range, still not sure what exactly it might be though. 
- `"D0312F"` is now 50, previously it was always 0. Don't know what it might refer to. Apparently the device has a brightness sensor. Since i last tested it in the evening and now when there's light, maybe this is related? However, i did have lights on… so no idea. It's also possible that it is referring to AQI. The app currently shows a value of "51/500" for my "area", but not as a measurement by my device. Fun side note: The Allergy Risk in my region is currently "/null". Fascinating.
- `"D0520D"` has gone from 720 to 719 and 718, indicating it is slowly going down with time, probably.

Most importantly: **The air quality indicator on the device is now purple. Previously it was blue.**

![image](https://user-images.githubusercontent.com/3868507/237018449-535b1444-6f16-40d9-a486-c39d06d8aeb2.png)

Looking at the app and the manual, the purple can't be caused by the PM2.5 concentration, since at these levels it should be rather dark blue and i'd say it's not. So it has to be another sensor causing it, and the unknown sensor `"D03120"` would be a likely culprit. 

![image](https://user-images.githubusercontent.com/3868507/237019179-581e2e12-3de5-4a06-a8b5-0d2bc81d712d.png)
![image](https://user-images.githubusercontent.com/3868507/237019258-fc9bb866-68f1-4b18-8100-d392a57ab62b.png)
![image](https://user-images.githubusercontent.com/3868507/237019286-b92c3a33-2dbe-4231-ab60-2c6139e633c2.png)

Most notably, the manual that came with the device does mention IAI measurements being only available through the app. Nowever, the app does not show it and the more extensive version of the manual you can find as a PDF on Philips' website also no longer mentions IAI. But, even in the section where it is mentioned it seems like that would explain a dark blue and not a purple. Hm.

I have turned the magic Auto+ AI mode. You can still set the fan speed while it is on and it doesn't turn off. That made me suspicious, maybe it is only active if the fan speed is set to auto? Turns out: Yes. If you turn on Auto+ AI mode, the app swaps the "Auto mode" with "Auto+ Mode". `"D0310C"` still refers to Auto+ mode as `0` while it's active, just as with Auto mode. 

The app is also now proudly telling me when Auto+ mode is "active" with statements like "Auto+ mode was activated 2 times to keep your air clean and healthy." and "Auto+ mode has increasing filtering and fan speed. At 9:21, Pm2.5 increased to 15ug/m³. Will lower in minutes.". Apparently the app keeps track of when the mode Auto+ mode "engages" during Auto mode and tells me when and what it did. There's no indication of this in the JSON dump.

While writing this, the indicator has turned blue. I would say it's the bright blue, not the intermediate blue. `"D03120"` is now at 2 and `"D03221"` is at 7. So it's unclear which one of the two caused it. Maybe the purple was actually that dark-purple-blueish color? It would fit with both PM2.5 and IAI according to the paper manual then. Of course

Anyway, this brings us to this:

| Key	        | Example Value	        | Seen values	| Description
| ------------- | ------------- | ------------- | ------------------------| 
| "D01102"	| 5		| always 5 	|
| "D01S03"	| "Living Room" |               | Device Name given by User |
| "D01S04"	| "Pegasus" 	|               | Model name, maybe?
| "D01S05"	| "AMF765/10" 	|               | Model number
| "D01107"	| 0		| always 0 	|
| "D01108"	| 3		| always 3 	|
| "D01109"	| 1		| always 1 	|
| "D0110A"	| 0		| always 0 	|
| "D0110B"	| 7		| always 7 	|
| "D0110C"	| 3		| always 3 	|
| "D0110F"	| 3		| always 3 	|
| "D01S12"	| "0.8.6"	| 0.7.3, 0.8.6	| Device Firmware Version |
| "D01213"	| 0		| always 0 	|
| "D01R22"	| "0.1.5"	| always 0.1.5 	|
| "D01223"	| 0		| always 0 	|
| "D01R32"	| "0.3.0"	| always 0.3.0 	|
| "D01233"	| 0		| always 0 	|
| "ProductId"	| "5537ddae116411eda7a60217247a73aa" |
| "DeviceId"	| "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" |
| "MCUBoot"	| true		| always true 	|
| "Runtime"	| 701713	| 		| counting up, milliseconds |
| "rssi"	| -34		| 
| "wifilog"	| false		| always false 	|
| "free_memory"	| 47096		| 
| "WifiVersion"	| "AWS_Philips_AIR@78" | `@72.6`, `@78` | WiFi Firmware Version |
| "StatusType"	| "status"	| "status", "control" |
| "ConnectType"	| "Online"	| "Online", 	|
| "D03102"	| 1		| 0, 1		| Power status (0=standby, 1=on) |
| "D03103"	| 0		| 0, 1		| Child lock (0=off, 1=on) |
| "D03106"	| 0		| 0, 1 		| no idea what made it go from 0 to 1
| "D0310A"	| 2		| 1, 2		| Mode (1=fan, 2=circulation) (opens/closes extra vent) |
| "D0310C"	| 5		| 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 17, 18 | Fan setting (0=auto/auto+ mode, 1-10=manual speed, 17=sleep mode, 18=turbo mode) – 0 during standby |
| "D0310D"	| 5		| 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 18 | Fan actual (1-10=speed, 18=turbo) – 0 during standby |
| "D0310E"	| 25		| always 25	| (not temperature)
| "D0320F"	| 0		| 0, 45, 50, 65, …, 350	| fan oscillation degrees |
| "D03110"	| 0		| 0-13		| poweroff timer (0=off, 1=0.5h, 2=1h, 3=2h, 4=3h, ..., 13=12h) |
| "D03113"	| 0		| always 0 	|
| "D03114"	| 0		| always 0 	|
| "D0311F"	| 0		| always 0 	|
| "D03120"	| 2		| 0, 1, 2, 4, 5	| unknown sensor value, IAI maybe? app doesn't show it. – 0 during standby if sensors are off |
| "D03221"	| 7		| 2-22		| PM2.5 – 0 during standby if sensors are off |
| "D03122"	| 1		| always 1 	|
| "D03123"	| 50		| always 50 	|
| "D03224"	| 236		| 238-249	| temperature in °C ×10 |
| "D03125"	| 57		| 53-57		| relative humidity % |
| "D0312A"	| 1		| always 1 	|
| "D0312B"	| 1		| always 1 	|
| "D0312C"	| 4		| always 4 	|
| "D0312D"	| 0		| 0, 10, …, 100	| display brightness (0=off after a few seconds) |
| "D0312E"	| 30		| always 30 	|
| "D0312F"	| 0		| 0, 50		| first 0, now 50. unclear what it is
| "D03130"	| 100		| 0, 100	| beep volume, app only allows turning off and on, not sure if intermediate values are possible |
| "D03131"	| 0		| always 0 	|
| "D03134"	| 0		| 0, 1		| sensors active during standby (0=no, 1=yes) |
| "D03240"	| 0		| always 0 	|
| "D03180"	| 0		| 0, 1		| Replace Auto Mode with Auto+ Ai Mode (0=no, 1=yes) |
| "D03182"	| 1		| N/A, 1	| Only introduced in an update, wasn't there before, since then always 1 |
| "D03R81"	| "[endless screaming in base64]" |
| "D05102"	| 0		| always 0	|
| "D05207"	| 720		| always 720	|
| "D05408"	| 14400		| always 14400	|
| "D0520D"	| 720		| 718-720	| seems to continuously slowly count down
| "D0540E"	| 14400		| always 14400	|

**Comment by codingcatgirl** [comment]
The unresponsiveness over JSON couuld be fixed by unplugging the device and plugging it back in, btw. Hm.

Just saw the device cycle through it's colors during bootup and can confirm the color it displayed was level 2, not level 1.

**Comment by codingcatgirl** [comment]
Update: "D0520D" is the remaining time until the filter needs to be cleaned. It has reached 713 now and the app displays it as 99%, which adds up for a maximum value of 720. The unit might be hours, which would also add up to 30 days, which would make sense.

The app also displays a second statistic for how much usage is left until the filter needs to be replaced. This one is still at 100% and has not changed. Not sure which value this might refer to, but i will eventually find out.

**Comment by codingcatgirl** [comment]
Some more observations, not much new, but: 
- D0312F is still 50. not gone back to 0.
- D0540E is counting downwards, but at an unclear rate. While D0520D definitely was hours, it has now gone down 216, and D0540E has gone down 144. It defintely refers to filter replace lifetime.
- D05207 and D05408 refer to the maximum value for D0520D and D0540E respectively

Which brings us to this, and i think there's probably not going to be much more to find: 


| Key	        | Example Value	        | Seen values	| Description
| ------------- | ------------- | ------------- | ------------------------| 
| "D01102"	| 5		| always 5 	|
| "D01S03"	| "Living Room" |               | Device Name given by User |
| "D01S04"	| "Pegasus" 	|               | Model name, maybe?
| "D01S05"	| "AMF765/10" 	|               | Model number
| "D01107"	| 0		| always 0 	|
| "D01108"	| 3		| always 3 	|
| "D01109"	| 1		| always 1 	|
| "D0110A"	| 0		| always 0 	|
| "D0110B"	| 7		| always 7 	|
| "D0110C"	| 3		| always 3 	|
| "D0110F"	| 3		| always 3 	|
| "D01S12"	| "0.8.6"	| 0.7.3, 0.8.6	| Device Firmware Version |
| "D01213"	| 0		| always 0 	|
| "D01R22"	| "0.1.5"	| always 0.1.5 	|
| "D01223"	| 0		| always 0 	|
| "D01R32"	| "0.3.0"	| always 0.3.0 	|
| "D01233"	| 0		| always 0 	|
| "ProductId"	| "5537ddae116411eda7a60217247a73aa" |
| "DeviceId"	| "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" |
| "MCUBoot"	| true		| always true 	|
| "Runtime"	| 701713	| 		| counting up, milliseconds |
| "rssi"	| -34		| 
| "wifilog"	| false		| always false 	|
| "free_memory"	| 47096		| 
| "WifiVersion"	| "AWS_Philips_AIR@78" | `@72.6`, `@78` | WiFi Firmware Version |
| "StatusType"	| "status"	| "status", "control" |
| "ConnectType"	| "Online"	| "Online", 	|
| "D03102"	| 1		| 0, 1		| Power status (0=standby, 1=on) |
| "D03103"	| 0		| 0, 1		| Child lock (0=off, 1=on) |
| "D03106"	| 0		| 0, 1 		| no idea what made it go from 0 to 1
| "D0310A"	| 2		| 1, 2		| Mode (1=fan, 2=circulation) (opens/closes extra vent) |
| "D0310C"	| 5		| 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 17, 18 | Fan setting (0=auto/auto+ mode, 1-10=manual speed, 17=sleep mode, 18=turbo mode) – 0 during standby |
| "D0310D"	| 5		| 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 18 | Fan actual (1-10=speed, 18=turbo) – 0 during standby |
| "D0310E"	| 25		| always 25	| (not temperature)
| "D0320F"	| 0		| 0, 45, 50, 65, …, 350	| fan oscillation degrees |
| "D03110"	| 0		| 0-13		| poweroff timer (0=off, 1=0.5h, 2=1h, 3=2h, 4=3h, ..., 13=12h) |
| "D03113"	| 0		| always 0 	|
| "D03114"	| 0		| always 0 	|
| "D0311F"	| 0		| always 0 	|
| "D03120"	| 2		| 0, 1, 2, 4, 5	| unknown sensor value, IAI maybe? app doesn't show it. – 0 during standby if sensors are off |
| "D03221"	| 7		| 2-22		| PM2.5 – 0 during standby if sensors are off |
| "D03122"	| 1		| always 1 	|
| "D03123"	| 50		| always 50 	|
| "D03224"	| 236		| 238-249	| temperature in °C ×10 |
| "D03125"	| 57		| 53-57		| relative humidity % |
| "D0312A"	| 1		| always 1 	|
| "D0312B"	| 1		| always 1 	|
| "D0312C"	| 4		| always 4 	|
| "D0312D"	| 0		| 0, 10, …, 100	| display brightness (0=off after a few seconds) |
| "D0312E"	| 30		| always 30 	|
| "D0312F"	| 0		| 0, 50		| first 0, now 50. unclear what it is
| "D03130"	| 100		| 0, 100	| beep volume, app only allows turning off and on, not sure if intermediate values are possible |
| "D03131"	| 0		| always 0 	|
| "D03134"	| 0		| 0, 1		| sensors active during standby (0=no, 1=yes) |
| "D03240"	| 0		| always 0 	|
| "D03180"	| 0		| 0, 1		| Replace Auto Mode with Auto+ Ai Mode (0=no, 1=yes) |
| "D03182"	| 1		| N/A, 1	| Only introduced in an update, wasn't there before, since then always 1 |
| "D03R81"	| "[endless screaming in base64]" |
| "D05102"	| 0		| always 0	|
| "D05207"	| 720		| always 720	| how many hours of use the filter offers in total until it needs to be **cleaned**
| "D05408"	| 14400		| always 14400	| how many [unit?] of use the filter offers in total until it needs to be **replaced**
| "D0520D"	| 720		| 504-720	| how many hours of use remain until the the filter needs to be **cleaned** (0-`D05207`)
| "D0540E"	| 14400		| 14256-14400	| how many [unit?] of use remain until the the filter needs to be **replaced** (0-`D05408`)

**Comment by tjorim** [comment]
I'm considering buying the Philips AMF765/10 as well. I don't think (preliminary) support for this model is in the releases yet?

**Comment by kongo09** [comment]
I still need to find time to look at this but with all the data above I feel cautiously optimistic

**Comment by tjorim** [comment]
Great, I just got one and made this dump:

|              |                                                                                            |
|--------------|--------------------------------------------------------------------------------------------|
| D01102       | 5                                                                                          |
| D01S03       | My Device                                                                                  |
| D01S04       | Pegasus                                                                                    |
| D01S05       | AMF765/10                                                                                  |
| D01107       | 0                                                                                          |
| D01108       | 3                                                                                          |
| D01109       | 1                                                                                          |
| D0110A       | 0                                                                                          |
| D0110B       | 7                                                                                          |
| D0110C       | 3                                                                                          |
| D0110F       | 3                                                                                          |
| D01S12       | 0\.8\.6                                                                                    |
| D01213       | 0                                                                                          |
| D01R22       | 0\.1\.5                                                                                    |
| D01223       | 0                                                                                          |
| D01R32       | 0\.3\.0                                                                                    |
| D01233       | 0                                                                                          |
| ProductId    | 5537ddae116411eda7a60217247a73aa                                                           |
| DeviceId     | d7a16a6609dc11ee8baedd9096030e84                                                           |
| MCUBoot      | true                                                                                       |
| Runtime      | 1723520                                                                                    |
| rssi         | \-52                                                                                       |
| wifilog      | false                                                                                      |
| free\_memory | 40776                                                                                      |
| WifiVersion  | AWS\_Philips\_AIR@78                                                                       |
| StatusType   | status                                                                                     |
| ConnectType  | Online                                                                                     |
| D03102       | 1                                                                                          |
| D03103       | 0                                                                                          |
| D03106       | 0                                                                                          |
| D0310A       | 1                                                                                          |
| D0310C       | 10                                                                                         |
| D0310D       | 10                                                                                         |
| D0310E       | 25                                                                                         |
| D0320F       | 90                                                                                         |
| D03110       | 0                                                                                          |
| D03113       | 0                                                                                          |
| D03114       | 0                                                                                          |
| D0311F       | 0                                                                                          |
| D03120       | 1                                                                                          |
| D03221       | 3                                                                                          |
| D03122       | 1                                                                                          |
| D03123       | 50                                                                                         |
| D03224       | 263                                                                                        |
| D03125       | 42                                                                                         |
| D0312A       | 1                                                                                          |
| D0312B       | 1                                                                                          |
| D0312C       | 4                                                                                          |
| D0312D       | 100                                                                                        |
| D0312E       | 30                                                                                         |
| D0312F       | 0                                                                                          |
| D03130       | 100                                                                                        |
| D03131       | 0                                                                                          |
| D03134       | 0                                                                                          |
| D03240       | 0                                                                                          |
| D03180       | 1                                                                                          |
| D03182       | 1                                                                                          |
| D03R81       | AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA== |
| D05102       | 0                                                                                          |
| D05207       | 720                                                                                        |
| D05408       | 14400                                                                                      |
| D0520D       | 719                                                                                        |
| D0540E       | 14400                                                                                      |

PS: any way to sponsor your work?

**Comment by codingcatgirl** [comment]
It's been a month, i just had to clean the filter for the first time, and i have some new additional observations :)

- Auto+ Mode only replaces Auto mode when the Device is set to circulation, not if it's set to fan. So, Auto+ mode is only active if: `D03180` is 1, `D0310C` is 0 and `D0310A` is 2. The app does not seem to be aware of this behavior, but you can tell from how the device operates (thresholds for speeds seem to be different), and it also uses the LED indicator for auto mode instead of "app-only feature" mode.
- The value of `D03240` has suddenly jumped to `49153` for me. I assume this happened when the filter was set to "needs cleaning". But, even resetting the countdown doesn't reset this value back to 0. Maybe it's some kind of statistic?

Which brings us to this:

| Key	        | Example Value	        | Seen values	| Description
| ------------- | ------------- | ------------- | ------------------------| 
| "D01102"	| 5		| always 5 	|
| "D01S03"	| "Living Room" |               | Device Name given by User |
| "D01S04"	| "Pegasus" 	|               | Model name, maybe?
| "D01S05"	| "AMF765/10" 	|               | Model number
| "D01107"	| 0		| always 0 	|
| "D01108"	| 3		| always 3 	|
| "D01109"	| 1		| always 1 	|
| "D0110A"	| 0		| always 0 	|
| "D0110B"	| 7		| always 7 	|
| "D0110C"	| 3		| always 3 	|
| "D0110F"	| 3		| always 3 	|
| "D01S12"	| "0.8.6"	| 0.7.3, 0.8.6	| Device Firmware Version |
| "D01213"	| 0		| always 0 	|
| "D01R22"	| "0.1.5"	| always 0.1.5 	|
| "D01223"	| 0		| always 0 	|
| "D01R32"	| "0.3.0"	| always 0.3.0 	|
| "D01233"	| 0		| always 0 	|
| "ProductId"	| "5537ddae116411eda7a60217247a73aa" |
| "DeviceId"	| "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" |
| "MCUBoot"	| true		| always true 	|
| "Runtime"	| 701713	| 		| counting up, milliseconds |
| "rssi"	| -34		| 
| "wifilog"	| false		| always false 	|
| "free_memory"	| 47096		| 
| "WifiVersion"	| "AWS_Philips_AIR@78" | `@72.6`, `@78` | WiFi Firmware Version |
| "StatusType"	| "status"	| "status", "control" |
| "ConnectType"	| "Online"	| "Online", 	|
| "D03102"	| 1		| 0, 1		| Power status (0=standby, 1=on) |
| "D03103"	| 0		| 0, 1		| Child lock (0=off, 1=on) |
| "D03106"	| 0		| 0, 1 		| no idea what made it go from 0 to 1
| "D0310A"	| 2		| 1, 2		| Mode (1=fan, 2=circulation) (opens/closes extra vent) |
| "D0310C"	| 5		| 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 17, 18 | Fan setting (0=auto/auto+ mode (auto+ if enabled and circulation), 1-10=manual speed, 17=sleep mode, 18=turbo mode) – 0 during standby |
| "D0310D"	| 5		| 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 18 | Fan actual (1-10=speed, 18=turbo) – 0 during standby |
| "D0310E"	| 25		| always 25	| (not temperature)
| "D0320F"	| 0		| 0, 45, 50, 65, …, 350	| fan oscillation degrees |
| "D03110"	| 0		| 0-13		| poweroff timer (0=off, 1=0.5h, 2=1h, 3=2h, 4=3h, ..., 13=12h) |
| "D03113"	| 0		| always 0 	|
| "D03114"	| 0		| always 0 	|
| "D0311F"	| 0		| always 0 	|
| "D03120"	| 2		| 0, 1, 2, 4, 5	| unknown sensor value, IAI maybe? app doesn't show it. – 0 during standby if sensors are off |
| "D03221"	| 7		| 2-22		| PM2.5 – 0 during standby if sensors are off |
| "D03122"	| 1		| always 1 	|
| "D03123"	| 50		| always 50 	|
| "D03224"	| 236		| 238-249	| temperature in °C ×10 |
| "D03125"	| 57		| 53-57		| relative humidity % |
| "D0312A"	| 1		| always 1 	|
| "D0312B"	| 1		| always 1 	|
| "D0312C"	| 4		| always 4 	|
| "D0312D"	| 0		| 0, 10, …, 100	| display brightness (0=off after a few seconds) |
| "D0312E"	| 30		| always 30 	|
| "D0312F"	| 0		| 0, 50		| first 0, now 50. unclear what it is
| "D03130"	| 100		| 0, 100	| beep volume, app only allows turning off and on, not sure if intermediate values are possible |
| "D03131"	| 0		| always 0 	|
| "D03134"	| 0		| 0, 1		| sensors active during standby (0=no, 1=yes) |
| "D03240"	| 0		| 0, 49135 	| zero until filter first needed to be cleaned, reset of filter timer does not reset this value though, statistic maybe? |
| "D03180"	| 0		| 0, 1		| Replace Auto Mode with Auto+ Ai Mode (if circulation) (0=no, 1=yes) |
| "D03182"	| 1		| N/A, 1	| Only introduced in an update, wasn't there before, since then always 1 |
| "D03R81"	| "[endless screaming in base64]" |
| "D05102"	| 0		| always 0	|
| "D05207"	| 720		| always 720	| how many hours of use the filter offers in total until it needs to be **cleaned**
| "D05408"	| 14400		| always 14400	| how many [unit?] of use the filter offers in total until it needs to be **replaced**
| "D0520D"	| 720		| 0-720	| how many hours of use remain until the the filter needs to be **cleaned** (0-`D05207`)
| "D0540E"	| 14400		| 13680-14400	| how many [unit?] of use remain until the the filter needs to be **replaced** (0-`D05408`)

**Comment by suxus-zz** [comment]
Hello
I try to install the beta v0.11.4-beta765.2 for my new AMF765, but i can't find in the repository list. How can install this beta for my new AMF765?
Thank you very much
<img width="657" alt="install" src="https://github.com/kongo09/philips-airpurifier-coap/assets/17749132/ee44b778-97e9-4a44-a0f6-9bb043a9d6e2">

**Comment by kongo09** [comment]
I need to find some time to do a new beta

**Comment by suxus-zz** [comment]
@kongo09 is it possible that i can download the old beta v0.11.4-beta765.2 and replace files manually? If yes witch file must replaced?
My idea is i install the v0.15.0 and replace the files from the beta if this is even possible.
regards

**Comment by suxus-zz** [comment]
Hello, do you already know when a new version is planned to support the AMF765?
Thank you very much

**Comment by tjorim** [comment]
I guess people get busy with their offline life. Let me know if there's a way we can support you @kongo09 !

**Comment by tjorim** [comment]
I guess people get busy with their offline life. Are we missing any other keys? Let me know if there's a way we can support you @kongo09

**Comment by kongo09** [comment]
Thanks for all the detailed information. I'll try and put something together. Apologies that I didn't find much time recently.

**Comment by codingcatgirl** [comment]
Exciting news! Looking forward to it. Take your time, this is still a free project. Hope you have a good new year!

**Comment by kongo09** [comment]
> PS: any way to sponsor your work?

I've just put up a "buy me a coffee button" in the README :-)

But seriously, I'll tackle this device next. It's a completely different schema but with all the great data that you collected, it should be possible to put something together.

One complexity will be that some settings might have to be set together to work. I assume that's the case for the Auto+ mode for example, which also needs the Auto mode to be set. But I've seen this also for speed modes being the case, etc... We might have to experiment here and I need you guys to try and report back what happens.

I'm not yet clear about your lights: is there just one type of light that's dimmable?

Is the circulation function always in addition to the purifier function, or are these two alternatives, i.e. do I do a switch "circulation on/off" or do I do a select "circulation vs. purification"? And do the manual speed modes work in both settings?

What is the range of rotation degree angles for the oscillation? The manual says: off, 45, 90, 180, 270, 350. But the values you've given above are different, e.g. there is a 50 in the table above?

The manual also talks about an "app mode". What is that? For all the other devices, the app works in parallel to the device buttons and it doesn't matter which one you set, it stays always in sync. So I don't really understand what an app mode would be?

The manual also mentions a light sensor. Could someone please make an experiment and without changing anything else get a json dump of the light in the room being on and then one json dump with the light in the room being off. That should show us, if one of the value gives us a light sensor reading.

As for the filter time sensor: If this device is similar to the other devices, you would have two sensors with four json values. The first two would give the maximum filter time in hours and the second two would run down over time and then eventually lead to the warning. Typically, the count down values should go down in sync from 720 and 14400 respectively. The percentages would obviously be different. I'll go ahead an implement this just like in the other Philips models and then you can see if it makes any sense in practice.

**Comment by codingcatgirl** [comment]
> I'm not yet clear about your lights: is there just one type of light that's dimmable?

The entire display dims as a whole. It looks very similar to the displays on other purifiers.

> What is the range of rotation degree angles for the oscillation? The manual says: off, 45, 90, 180, 270, 350. But the values you've given above are different, e.g. there is a 50 in the table above?

As given in my table, it can be 0, 45, 50, 65, …, 350, by which i mean anything from 45 to 350 in steps of 5, as well as 0.

> Is the circulation function always in addition to the purifier function, or are these two alternatives, i.e. do I do a switch "circulation on/off" or do I do a select "circulation vs. purification"? And do the manual speed modes work in both settings?

The circulation mode opens an additional extra vent in the back of the device. It takes like 20 seconds to open and close (you can hear the servo). It does not affect any operational modes, it's still a fan then but far weaker since the air goes in two directions. I can't test it right now but if memory servers me right, Auto+ mode requires circulation mode. Meaning if you disable circulation, auto refers to auto, not auto+. If you turn on auto in the app while "replace auto with auto+" is enabled, it automatically disables the circulation mode. 

> The manual also talks about an "app mode". What is that?

That's just an additional icon on the display they use when the fan is in a mode that you can only enable using the app (and not using the remove). Presently, that's only Auto+ mode.

> The manual also mentions a light sensor.

Can't test this right now unfortunately. 

> As for the filter time sensor: 

Sounds right to me

**Comment by kongo09** [comment]
I've done a first beta: v0.17-amfcx-beta.1

It can't do much but it should talk to the device, allow you to switch it on and off and also make the lights available. The device entity should also have some few attributes.

Can you please test and let me know if that works?

**Comment by grssll** [comment]
@kongo09 
I have tried to add and getting "uknown error occured"
When you adding do you type just the ip address or something else?
Thanks
Just noticed that my model is 765/70 if it makes a difference. They look the same

**Comment by grssll** [comment]
@kongo09 
On one of your previous version I was able to see the purifier but not to control it

**Comment by tjorim** [comment]
Device does not get discovered, manually connecting does not work. Please see debug logs attached:
[home-assistant_2024-01-10T01-12-33.629Z.log](https://github.com/kongo09/philips-airpurifier-coap/files/13881724/home-assistant_2024-01-10T01-12-33.629Z.log)

**Comment by kongo09** [comment]
Thanks. I don't know that the `coap` errors mean, that's in the protocol / library. But it seems like there an initial connection and then a bug in my code.

When you test the next beta, please enable debug logging. Please try: v0.17-amfcx-beta.2

**Comment by grssll** [comment]
@kongo09 
Beta 2 connected to the AMF but no devices or entities available.

**Comment by kongo09** [comment]
Can you please share debug logs

**Comment by tjorim** [comment]
At first, I got similar logs with beta 2. I went ahead and powercycled the air purifier. Things changed.
Here is a new log attached. At first it seems to discover it but gets abored because a timeout.
Afterwards I manually connected again, device gets created but no entities.
[home-assistant_2024-01-10T12-02-36.418Z.log](https://github.com/kongo09/philips-airpurifier-coap/files/13887416/home-assistant_2024-01-10T12-02-36.418Z.log)

These errors stand out:

<details><summary>Error while setting up philips_airpurifier_coap platform for fan</summary>
<p>
Logger: homeassistant.components.fan
Source: helpers/entity_platform.py:360
Integration: Fan (documentation, issues)
First occurred: 13:02:16 (1 occurrences)
Last logged: 13:02:16

Error while setting up philips_airpurifier_coap platform for fan
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 360, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/philips_airpurifier_coap/fan.py", line 34, in async_setup_entry
    device = model_class(
             ^^^^^^^^^^^^
  File "/config/custom_components/philips_airpurifier_coap/philips.py", line 290, in __init__
    super().__init__(coordinator, model, name)
  File "/config/custom_components/philips_airpurifier_coap/philips.py", line 250, in __init__
    super().__init__(coordinator)
  File "/config/custom_components/philips_airpurifier_coap/philips.py", line 183, in __init__
    self._name = list(
                 ^^^^^
IndexError: list index out of range
</p>
</details> 

<details><summary>Error while setting up philips_airpurifier_coap platform for sensor</summary>
<p>
Logger: homeassistant.components.sensor
Source: helpers/entity_platform.py:360
Integration: Sensor (documentation, issues)
First occurred: 13:02:16 (1 occurrences)
Last logged: 13:02:16

Error while setting up philips_airpurifier_coap platform for sensor
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 360, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/philips_airpurifier_coap/sensor.py", line 58, in async_setup_entry
    sensors.append(PhilipsSensor(coordinator, name, model, sensor))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/philips_airpurifier_coap/sensor.py", line 81, in __init__
    super().__init__(coordinator)
  File "/config/custom_components/philips_airpurifier_coap/philips.py", line 183, in __init__
    self._name = list(
                 ^^^^^
IndexError: list index out of range
</p>
</details> 

<details><summary>Error while setting up philips_airpurifier_coap platform for light</summary>
<p>
Logger: homeassistant.components.light
Source: helpers/entity_platform.py:360
Integration: Light (documentation, issues)
First occurred: 13:02:16 (1 occurrences)
Last logged: 13:02:16

Error while setting up philips_airpurifier_coap platform for light
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 360, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/philips_airpurifier_coap/light.py", line 65, in async_setup_entry
    lights.append(PhilipsLight(coordinator, name, model, light))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/philips_airpurifier_coap/light.py", line 82, in __init__
    super().__init__(coordinator)
  File "/config/custom_components/philips_airpurifier_coap/philips.py", line 183, in __init__
    self._name = list(
                 ^^^^^
IndexError: list index out of range
</p>
</details> 

After a few more minutes there are some more debug logs:
[home-assistant_2024-01-10T12-11-52.110Z.log](https://github.com/kongo09/philips-airpurifier-coap/files/13887455/home-assistant_2024-01-10T12-11-52.110Z.log)

**Comment by kongo09** [comment]
Ok, there are two flows, one with dhcp based autodiscovery. It often fails so I didn't put the new models into this. I've fixed that now in [v0.17-amfcx-beta.3](https://github.com/kongo09/philips-airpurifier-coap/releases/tag/v0.17-amfcx-beta.3) So please try again.

The manual flow was where I had the first code pieces. The it looks like something is happening, but we're not quite there yet. Will look into beta 4 later.

**Comment by kongo09** [comment]
v0.17-amfcx-beta.4

**Comment by tjorim** [comment]
Needed to powercycle the air purifier to show up. Here are the discovery logs (beta 3):
```
2024-01-10 15:52:54.167 DEBUG (MainThread) [custom_components.philips_airpurifier_coap.config_flow] async_step_dhcp: called, found: DhcpServiceInfo(ip='192.168.0.114', hostname='mxchip', macaddress='849dc27a6c6e')
2024-01-10 15:52:54.169 DEBUG (MainThread) [custom_components.philips_airpurifier_coap.config_flow] trying to configure host: 192.168.0.114
2024-01-10 15:52:54.221 DEBUG (MainThread) [aioairctrl.coap.client] syncing
2024-01-10 15:52:54.230 DEBUG (MainThread) [coap] Sending request - Token: c300, Remote: <UDP6EndpointAddress 192.168.0.114>
2024-01-10 15:52:54.231 DEBUG (MainThread) [coap] Sending message <aiocoap.Message at 0x7f2d3d30d0: NON POST (MID 35925, token c300) remote <UDP6EndpointAddress 192.168.0.114>, 1 option(s), 8 byte(s) payload>
2024-01-10 15:53:24.170 WARNING (MainThread) [custom_components.philips_airpurifier_coap.config_flow] Timeout, host 192.168.0.114 looks like a Philips AirPurifier but doesn't answer, aborting
```

**Comment by tjorim** [comment]
Basic functionality works! I have 3 entities:
![image](https://github.com/kongo09/philips-airpurifier-coap/assets/1294876/4cae770f-a7c9-4658-8df4-876e542cfb86)
I can turn the device on and off, but presets are missing. The display 'backlight' switch and wifi strength sensor also work.
Here are some more debug logs in case you need them (just playing with the device, both through HA, remote and app):
[home-assistant_2024-01-10T15-04-11.246Z.log](https://github.com/kongo09/philips-airpurifier-coap/files/13889456/home-assistant_2024-01-10T15-04-11.246Z.log)

**Comment by kongo09** [comment]
That's amazing! I didn't implement anything else, just on/off and display backlight. Please observe this and play around with this. Make sure the dimming works. Also make sure that this stays in sync with changes on the device and the app.

**Comment by tjorim** [comment]
From the few minutes I tried it kept up pretty good with external changes. Changing it in Home Assistant sometimes makes the switch bounce back. (edit: changes sometimes also takes a while to get reflected in HA)
Not sure what you mean by the dimming, right now the screen is a single switch on/off. I did notice that toggling the fan entity off and on again, turns on the display as well. (edit: scratch that last part, could not reproduce)

**Comment by kongo09** [comment]
With dimming I mean clicking on the display backlight button, see if there is a dimming in the interface and check if that does anything.

**Comment by tjorim** [comment]
Yes it works, I also enabled dimming: https://github.com/kongo09/philips-airpurifier-coap/pull/104

**Comment by grssll** [comment]
It works and reports back to HA.

**Comment by kongo09** [comment]
Ok, now with the basics working, I'll take the sensors next, as they are read only

**Comment by grssll** [comment]
@kongo09 ,
Any chance to fan speed change?
Thanks

**Comment by kongo09** [comment]
Next iteration with quite a few additions on the filter front: v0.17-amfcx-beta.5

**Comment by kongo09** [comment]
> Any chance to fan speed change?

we'll get there

**Comment by tjorim** [comment]
> Next iteration with quite a few additions on the filter front: v0.17-amfcx-beta.5

Beta 6 I got two errors:

<details><summary>Errors</summary>
<p>

```
Logger: homeassistant.loader
Source: loader.py:822
First occurred: 01:46:49 (1 occurrences)
Last logged: 01:46:49

Unexpected exception importing component custom_components.philips_airpurifier_coap
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/loader.py", line 822, in get_component
    ComponentProtocol, importlib.import_module(self.pkg_path)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/config/custom_components/philips_airpurifier_coap/__init__.py", line 18, in <module>
    from .const import (
  File "/config/custom_components/philips_airpurifier_coap/const.py", line 541, in <module>
    FanAttributes.TOTAL: PhilipsApi.NEW2_FILTER_NANOPROTECT_CLEAN_TOTAL,
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: type object 'PhilipsApi' has no attribute 'NEW2_FILTER_NANOPROTECT_CLEAN_TOTAL'
```

```
Logger: homeassistant.setup
Source: setup.py:251
First occurred: 01:46:49 (1 occurrences)
Last logged: 01:46:49

Setup failed for custom integration 'philips_airpurifier_coap': Unable to import component: Exception importing custom_components.philips_airpurifier_coap
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/loader.py", line 822, in get_component
    ComponentProtocol, importlib.import_module(self.pkg_path)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/config/custom_components/philips_airpurifier_coap/__init__.py", line 18, in <module>
    from .const import (
  File "/config/custom_components/philips_airpurifier_coap/const.py", line 541, in <module>
    FanAttributes.TOTAL: PhilipsApi.NEW2_FILTER_NANOPROTECT_CLEAN_TOTAL,
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: type object 'PhilipsApi' has no attribute 'NEW2_FILTER_NANOPROTECT_CLEAN_TOTAL'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/setup.py", line 251, in _async_setup_component
    component = integration.get_component()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/loader.py", line 830, in get_component
    raise ImportError(f"Exception importing {self.pkg_path}") from err
ImportError: Exception importing custom_components.philips_airpurifier_coap
```

</p>
</details> 

With my fixes in https://github.com/kongo09/philips-airpurifier-coap/pull/105 it seems to work.

**Comment by kongo09** [comment]
The next beta is out with some more stuff. Errors should be gone - thanks @tjorim 

Let me know how it goes.

**Comment by grssll** [comment]
Thanks. Much appreciated

**Comment by tjorim** [comment]
I now have two temperature sensors.
![Screenshot_20240112-002457.png](https://github.com/kongo09/philips-airpurifier-coap/assets/1294876/4deb3f82-c3f8-46a4-a686-c0d48be569ed)

I thought reading the commits the swing setting was now available but couldn't find it, only seems to be for the CX model.
Should the purification vs. circulation option be available for our model? Because I couldn't find it either.

Finally, one more warning.
```
Logger: homeassistant.components.sensor
Source: components/sensor/__init__.py:714
Integration: Sensor (documentation, issues)
First occurred: 00:19:25 (1 occurrences)
Last logged: 00:19:25

Entity sensor.air_purifier_temperature_2 (<class 'custom_components.philips_airpurifier_coap.sensor.PhilipsSensor'>) is using native unit of measurement '°C' which is not a valid unit for the device class ('speed') it is using; expected one of ['kn', 'in/h', 'km/h', 'mm/h', 'in/d', 'mph', 'mm/d', 'm/s', 'ft/s']; Please update your configuration if your entity is manually configured, otherwise create a bug report at https://github.com/kongo09/philips-airpurifier-coap/issues
```

**Comment by codingcatgirl** [comment]
Looking good so far for me as well. The Purifier seems to need an internet connection or it won't respond to the initial setup. I disabled it afterwards and can still control it now though.

![image](https://github.com/kongo09/philips-airpurifier-coap/assets/3868507/d183b094-1b61-418d-be88-0731996ce526)

The two temperature sensors are a bit confusing? Looking forward to more features :)

**Comment by codingcatgirl** [comment]
Never mind, the device seem to become unresponsive pretty quickly after internet access is cut. A workaround for this would be nice, though probably out of scope for this add-on?

Even after a power cycle the add-on seems to be unable to control device now. Hm.

**Comment by tjorim** [comment]
> Never mind, the device seem to become unresponsive pretty quickly after internet access is cut. A workaround for this would be nice, though probably out of scope for this add-on?
> 
> Even after a power cycle the add-on seems to be unable to control device now. Hm.

Just to avoid confusion: this is a (custom) integration, not an add-on.

**Comment by codingcatgirl** [comment]
Yes, obviously. Sorry.

**Comment by grssll** [comment]
Been testing for the last few days no problem whatsoever

**Comment by kongo09** [comment]
I see that icons are missing from the screenshot above. Can you please reload your frontend (potentially clear browser cache as a second step) to see if they show?

**Comment by kongo09** [comment]
The broken temp sensor should now be gone (and turned into an actual speed sensor). Please check.

> Should the purification vs. circulation option be available for our model?

Yes, if that is not showing up, something is broken. Anything in the debug logs?

**Comment by codingcatgirl** [comment]
> I see that icons are missing from the screenshot above. Can you please reload your frontend (potentially clear browser cache as a second step) to see if they show?

Yep, i have icons now :)

> The broken temp sensor should now be gone (and turned into an actual speed sensor). Please check.

It is indeed a fan sensor now.

> Yes, if that is not showing up, something is broken. Anything in the debug logs?

It's not showing up unfortunately and i also don't see anything in the debug logs. There are no debug logs at all, despite having enabled them. Is this normal?

What i just noticed is that the temperature updates… way too often when it's hoverign around a threshold. I just had a temperature update almost every second where it went back and forth by 0.1°C. IDK if this is maybe a new issue caused by this device… but wow this is going to really fill up the history database unnecessarily quickly. Could it maybe be filtered a bit?

I also noticed that the integration takes literal minutes to start up, seemingly slowing down the startup of other integrations.

**Comment by kongo09** [comment]
Startup time is a separate problem. I have no idea what that is.

Debug logs should be there, if you enable them.

Temperature update comes from the device. The integration just shows what the device is sending.

**Comment by kongo09** [comment]
two goodies in v0.17-amfcx-beta.10

Please check and share a screenshot of the device screen. Especially the oscillation selection is a bit a shot in the dark and a new control type across the Philips devices. The circulation selection should now be visible as well.

**Comment by tjorim** [comment]
The 'Function' selector does not work, it does not show the current state and picking any of them doesn't seem to actually change it. However it does change the function in the app to 'AI' (which is not an option actually). Also, purification/circulation sounds weird as the official app shows circulation/fan. Oscillation shows the correct value but the slider only goes from 0 to 100, the max is actually 350.

**Comment by kongo09** [comment]
Thanks, very helpful. I'll come to with a debug release to collect some logs

**Comment by kongo09** [comment]
The function selector is more challenging than I thought as there seem to be several settings to be relevant at the same time, as described here: https://github.com/kongo09/philips-airpurifier-coap/issues/63#issuecomment-1599255280

How does the app behave? What selections / switches would we need in HomeAssistant? And how are they interdependent?

**Comment by kongo09** [comment]
v0.17-amfcx-beta.12

**Comment by kongo09** [comment]
Can I please get some feedback about what works and what doesn't in the latest beta?

**Comment by grssll** [comment]
@kongo09 Sorry mate,
Everything seems to work correctly except fan/circulation mode.
Also, oscillation works correctly only after 45 degree steps. If you try to set the number before 45 it will always jump to 45.
Cheers

**Comment by kongo09** [comment]
> Also, oscillation works correctly only after 45 degree steps. If you try to set the number before 45 it will always jump to 45.

That is intended behaviour, as angles below 45 don't exist [according to the data above](https://github.com/kongo09/philips-airpurifier-coap/issues/63#issuecomment-1883868936). Or do they?

I've changed something in the circulation mode now, let's see if that works.

I've also integrated all the rest of the missing features including presets and fan speed. It's a bit of a shot in the dark, especially with the AI+ mode. It's simply a switch now setting only that single value, but maybe we have to set the function and auto mode at the same time. Please let me know how all of this looks and works.

Please try: v0.17-amfcx-beta.15

**Comment by grssll** [comment]
oscillation start at 30 then 35-40-45

**Comment by grssll** [comment]
@kongo09 
Beta 15: angles still dont work below 45. Speed works but the slider goes to OFF after changing speed.

**Comment by kongo09** [comment]
Angles is easy to fix, will do tomorrow, thanks for the info.

Speed feels strange. Does it set the correct speed, but then just moves the slider to OFF while the speed is still set? Or does the whole fan turn off then?

**Comment by grssll** [comment]
Sets the correct speed but then moves to off position. Fan still stays on the correct speed

**Comment by kongo09** [comment]
That feels like the actual value of the data field doesn't reflect the setting. Maybe that's why there is the second data field showing the fan speed. One for setting, one for reading?

I implemented the second one as sensor, but that's probably not correct. I'll try to combine the two, maybe that works. Will check tomorrow.

**Comment by grssll** [comment]
@kongo09 
Another interesting behaviour for the speed control.
Go to devices and speed control slider goes OFF.
Dashboard custom mushroom fan card correct behaviour when adjusting speed

**Comment by grssll** [comment]
see below
![Screenshot (9)](https://github.com/kongo09/philips-airpurifier-coap/assets/121098864/49604923-20bd-4a36-8993-f32a055a1f1c)
![Screenshot (10)](https://github.com/kongo09/philips-airpurifier-coap/assets/121098864/1fc7719c-bd9f-42e9-a6f2-2b2083f36d87)

**Comment by grssll** [comment]
Not updating from the app.
![Screenshot (11)](https://github.com/kongo09/philips-airpurifier-coap/assets/121098864/be98bed4-3be0-4fc1-a243-ee20408ff834)

**Comment by kongo09** [comment]
Please try v0.17-amfcx-beta.16

I've tried to fix the speed setting and also fixed the oscillation range. As a consequence, the speed sensor is removed as this should be covered by the speed setting. Let me know if it works and if anything is still missing.

**Comment by grssll** [comment]
@kongo09 
Beta 17. Speed slider still goes to off position after correctly setting the speed.
Oscillitaon works but the range below 30 could be removed as there no angles from 0-30.
If you set the angle at 10 or 20 it just jumps to 30 degree.
Speed sensor in beta 15 was working perfectly in sinc with the app.

**Comment by kongo09** [comment]
Removing the lower values for oscillation is not possible as we would lose the zero and then it cannot be turned off.

When you say speed in 15 was working, you mean it was also adjusting when you set the speed in the app? Above you said:

> Sets the correct speed but then moves to off position. Fan still stays on the correct speed

So I'm not really sure what the situation is. And forget mushroom cards. It needs to work in the device settings.

**Comment by grssll** [comment]
Beta 15:
Slider sets the correct speed but then moves to the off position.
Changes for the speed in the app does not reflect in the slider but correctly show speed in the speed sensor.
Beta17:
Fixed the oscilliation but not the speed slider in the device settings. Still goes down after setting the speed.
I hope its little bit clear
Thanks for your time

**Comment by kongo09** [comment]
So with respect to speed, beta 15 and 17 are the same?

**Comment by grssll** [comment]
> So with respect to speed, beta 15 and 17 are the same?

Yes, except speed sensor removed in 17

**Comment by kongo09** [comment]
I removed the sensor as this should work through the slider anyhow - in theory

Right now the slider tries to use the sensor values to figure out it's position, so there must be a bug

**Comment by kongo09** [comment]
please try v0.17-amfcx-beta.18

I fixed the pm2.5 sensor which should now show and also tried to fix preset and speed settings. Please let me know what happens

**Comment by kongo09** [comment]
So for the speed, I'm currently setting and reading `D0310D` and for the preset I'm setting and reading `D0310C`.

Can you confirm that setting works for both, speed and preset. The device reacts when you use the controls in HA. But reading doesn't work and the controls go to some neutral state. Is that true for both, speed and preset? What happens in preset when you change it on the device?

Can you look at the two data fields again and read them both before and after a preset change or a speed change. Does any other data field change at the same time? It feels like we're missing something.

**Comment by kongo09** [comment]
new beta: v0.17-amfcx-beta.19

I fixed and changed a few things. Hopefully everything is still fine here. Speed and presets still don't work.

**Comment by kongo09** [comment]
new beta: v0.17-amfcx-beta.20

another attempt at fixing speed and presets. Does this work?

**Comment by grssll** [comment]
Beta 20:
Presets, oscilliation works. Speed still the same goes to off position

**Comment by kongo09** [comment]
Preset works correctly? That's great and a step forward!

**Comment by kongo09** [comment]
> Beta 20: Presets, oscilliation works. Speed still the same goes to off position

Does the preset on HA also follow and change, if you set it in the Philips app?

**Comment by kongo09** [comment]
I had an idea about the presets and speed, please try: v0.17-amfcx-beta.21

**Comment by grssll** [comment]
> > Beta 20: Presets, oscilliation works. Speed still the same goes to off position
> 
> Does the preset on HA also follow and change, if you set it in the Philips app?

No, presets not following app.

**Comment by grssll** [comment]
Beta 21. Speed slider still the same goest to the off but ON/OFF button now works correctly.
I forgot to mention on/off button on the bottom of the slider wasnt working properly before.
In 21 it works to turn On or Off.

**Comment by kongo09** [comment]
I might have found something, please try beta.23

**Comment by grssll** [comment]
Still the same in 23. Everything works except the slider

**Comment by tjorim** [comment]
For me the slider works in beta 24. Attached some debug logs. Here is a short overview of what I did:

- was on auto + auto speed
- turned off in HA -> actually turned off
- turned on in HA -> actually turned on
- changed preset to auto -> actually changed to manual with speed 3
- changed preset to sleep -> actually changed to manual with speed 1
- changed preset to turbo -> actually went into turbo
- changed preset to auto -> saw it change to auto+ for a second in the app, then went to manual with speed 3 again
- changed speed to 50% -> speed changed to 5
- changed speed to 100% -> speed changed to 10
- changed speed to 10% -> speed changed to 1
- changed speed to 0% -> fan turned off
- changed preset to sleep -> did not pay attention to app
- changed preset to sleep again -> saw it change to sleep for a second in the app, then went to manual with speed 1 again
- tried changing the Display Backlight in HA -> no actual change but HA showed wrong
- tried changing the display backlight to 30% in the app -> not reflected in HA
- child lock OK, syncs right away, both ways
- beep OK, syncs right away, both ways
- standby sensors OK, syncs right away, both ways
- auto plus OK, syncs right away, both ways
- changed to auto+ preset in the app -> no changes in HA (could/should the active preset be shown in HA somewhere?)
- timer works OK
- oscillation works OK
- circulation vs fan function works OK (while it was on auto, changed between auto and auto+ as expected)

Just noticed the logs are quite long, sorry!
[home-assistant_philips_airpurifier_coap_2024-01-22T21-39-16.678Z.log](https://github.com/kongo09/philips-airpurifier-coap/files/14015725/home-assistant_philips_airpurifier_coap_2024-01-22T21-39-16.678Z.log)

**Comment by kongo09** [comment]
Thanks, I found some bugs, but not everything yet. But it feels like we're getting close. I've put quite some debug logging into the code, which is why the logs get long right now. That will be removed later once we fixed this.

**Comment by grssll** [comment]
Beta 24.
Slider now works and stays up and change of the speed reflects in the app.
Slider percentage sometimes works sometimes does not.
Speed change in the app does not reflect in the slider.

**Comment by grssll** [comment]
ON/OFF button stopped working correctly

**Comment by kongo09** [comment]
please try v0.17-amfcx-beta.25

I have a hypothesis that the auto+ mode might confuse me.

**Comment by grssll** [comment]
Seems to work now. I'll bit more testing and report tomorror.
Cheers

**Comment by tjorim** [comment]
I gave the beta 25 release a quick test. The issue now is that both auto and auto+ presets are available, even when the function is set to fan. Auto+ is just an 'upgrade' of the normal auto preset using 'AI'. It's a separate switch in the app (in the device settings) and worked just fine in beta 24. If you turn it off it goes back to the normal auto preset (next to sleep and turbo). It only works with the 'circulation' function, with the 'fan' function it shows as the normal auto mode. I think the switch could stay in the settings but personally I always leave it on. Here's a screenshot of the device setting:![Screenshot_20240122-235112.png](https://github.com/kongo09/philips-airpurifier-coap/assets/1294876/b98020f5-7a15-466a-8835-8b5579404ece)

**Comment by kongo09** [comment]
I can put it back into the switch. It seems like the two registers for the preset and speed settings can start to diverge if auto+ is enabled.

Right now I'm ignoring the second register and only set and read the first. I can turn the second register into a separate sensor if that makes any sense. I have no idea what the device is doing here with auto+ enabled and for example a speed set at the same time. Or is it just an alternative auto mode?

**Comment by kongo09** [comment]
I've put the Auto+ mode back into a switch. Can you please confirm if everything is working now in all modes and I've not forgotten any function?

Also, does the Turbo mode work as expected?

**Comment by grssll** [comment]
Beta 25 works as expected.

**Comment by kongo09** [comment]
That's great news. Thanks for all the help to get this going. I'll close this issue now. Please reopen if anything else pops up.

And thanks to everyone who got me a coffee!!! Much appreciated.
